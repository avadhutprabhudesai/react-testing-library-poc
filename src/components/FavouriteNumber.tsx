import React, { ChangeEventHandler, useState } from 'react';

type Props = {
  min?: number;
  max?: number;
};

export default function FavouriteNumber({ min = 1, max = 100 }: Props) {
  const [number, setNumber] = useState<number>(0);
  const [numberEntered, setNumberEntered] = useState<boolean>(false);
  const handleChange: ChangeEventHandler<HTMLInputElement> = (e) => {
    setNumber(+e.target.value);
    setNumberEntered(true);
  };
  const isValid = numberEntered && number >= min && number <= max;
  return (
    <>
      <label htmlFor="fav-number">Enter number</label>
      <input
        id="fav-number"
        type="number"
        value={number}
        onChange={handleChange}
      />
      {isValid ? <p role="alert">Number is valid</p> : null}
    </>
  );
}
