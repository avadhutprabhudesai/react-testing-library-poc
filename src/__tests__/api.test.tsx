import FavouriteNumber from 'components/FavouriteNumber';
import React from 'react';
import { render, screen } from 'test/test-utils';
import userEvent from '@testing-library/user-event';

describe('Testing React Testing Library API', () => {
  it('should render a component using custom render method', () => {
    render(<FavouriteNumber />);
    const input = screen.getByLabelText(/enter number/i);
    expect(input).toBeInTheDocument();
    expect(input).toHaveAttribute('id', 'fav-number');

    const message = screen.queryByRole(/alert/i);
    expect(message).not.toBeInTheDocument();
  });

  it('should re-render component with new props and update the dom accordingly', async () => {
    const { rerender } = render(<FavouriteNumber />);
    const user = userEvent.setup();
    const input = screen.getByLabelText(/enter number/i);
    await user.type(input, '123');

    expect(input).toBeInTheDocument();
    expect(input).toHaveDisplayValue('123');

    expect(screen.queryByRole(/alert/i)).not.toBeInTheDocument();

    rerender(<FavouriteNumber max={200} />);
    expect(screen.queryByRole(/alert/i)).toBeInTheDocument();
  });

  it('should capture rendered component in a DocumentFragment', async () => {
    const { asFragment } = render(<FavouriteNumber min={1} max={200} />);
    const initialRender = asFragment();

    const user = userEvent.setup();
    const input = screen.getByLabelText(/enter number/i);
    await user.type(input, '200');

    expect(initialRender).toMatchDiffSnapshot(asFragment());
  });
});
