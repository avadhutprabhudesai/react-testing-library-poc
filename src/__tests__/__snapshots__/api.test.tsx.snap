// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing React Testing Library API should capture rendered component in a DocumentFragment 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -7,6 +7,11 @@
    <input
      id=\\"fav-number\\"
      type=\\"number\\"
      value=\\"0\\"
    />
+   <p
+     role=\\"alert\\"
+   >
+     Number is valid
+   </p>
  </DocumentFragment>"
`;
